sql/core/src/main/scala/org/apache/spark/sql/execution/local/HashJoinNode.scala

```diff
@@ -73,10 +60,21 @@ case class HashJoinNode(
     }
   }
 
+  /** Sets the HashedRelation used by this node. */
+  protected def withHashedRelation(hashedRelation: HashedRelation): Unit = {
+    hashed = hashedRelation
+  }
+
+  /**
+   * For nodes that extends this, they can use doOpen to add operations needed in the open method.
+   * The implementation of this method should invoke its children's open methods.
+   */
+  protected def doOpen(): Unit
+
   override def open(): Unit = {
-    buildNode.open()
-    hashed = HashedRelation(buildNode, buildSideKeyGenerator)
-    streamedNode.open()
+    // First, call doOpen to invoke custom operations for a node.
+    doOpen()
+    // Second, initialize common internal states.
```

Comment:
These comments still seem unnecessary :). I would remove them but feel free to keep them if you prefer.